### ISSUE TITLE
Error loading Tensorflow model

### ISSUE THREAD

**Issue by lukealexmiller** [vespa-engine/vespa#9662]
I'm experiencing an error when loading a TF model and deploying the Vespa app:
```
Uploading application '/app/target/application.zip' using http://localhost:19071/application/v2/tenant/default/session
Session 2 for tenant 'default' created.
Preparing session 2 using http://localhost:19071/application/v2/tenant/default/session/2/prepared
Request failed. HTTP status code: 400
Invalid application package: default.default: Error loading model: Could not import TensorFlow model from directory '/opt/vespa/var/db/vespa/config_server/serverdb/tenants/default/sessions/2/.preprocessed/models/factorization_machine/saved': Output of 'dense_out/Sigmoid' has no type.
```
This also occurs if attempt to use a different layer as the output.

The model is trained in Keras, which I would assume to not be a problem as it's just an API to the underlying TF implementation.

My code for exporting the model is as used in: https://github.com/vespa-engine/vespa/blob/b3f6782b32990164ea703c7bcb44085cdb8c4fdb/model-integration/src/test/models/tensorflow/mnist/simple_mnist.py except for the I/O layer dicts.

The saved model and weights can be found on GDrive: https://drive.google.com/drive/folders/1kyvW88Y2_6RzQBQUVd8EYycFTvRx_vt4?usp=sharing

Thanks in advance

**Comment by lesters** [comment]
Thanks @lukealexmiller. I see a few import warnings here where the low level operations `VarHandleOp`, `ReadVariableOp`, `Pack`, `StridedSlice` and `ResourceGather` are not implemented in Vespa yet:

```
embedding_product_slug/embeddings: operation 'VarHandleOp' is currently not implemented
embedding_product_slug/embedding_lookup: operation 'ResourceGather' is currently not implemented
flatten/strided_slice: operation 'StridedSlice' is currently not implemented
flatten/Reshape/shape: operation 'Pack' is currently not implemented
embedding_user_guid/embeddings: operation 'VarHandleOp' is currently not implemented
embedding_user_guid/embedding_lookup: operation 'ResourceGather' is currently not implemented
flatten_1/strided_slice: operation 'StridedSlice' is currently not implemented
flatten_1/Reshape/shape: operation 'Pack' is currently not implemented
batch_normalization_v1/moving_variance: operation 'VarHandleOp' is currently not implemented
batch_normalization_v1/cond_1/ReadVariableOp: operation 'ReadVariableOp' is currently not implemented
batch_normalization_v1/gamma: operation 'VarHandleOp' is currently not implemented
batch_normalization_v1/batchnorm/mul/ReadVariableOp: operation 'ReadVariableOp' is currently not implemented
batch_normalization_v1/beta: operation 'VarHandleOp' is currently not implemented
batch_normalization_v1/batchnorm/ReadVariableOp: operation 'ReadVariableOp' is currently not implemented
batch_normalization_v1/moving_mean: operation 'VarHandleOp' is currently not implemented
batch_normalization_v1/cond/ReadVariableOp: operation 'ReadVariableOp' is currently not implemented
dense_out/kernel: operation 'VarHandleOp' is currently not implemented
dense_out/MatMul/ReadVariableOp: operation 'ReadVariableOp' is currently not implemented
dense_out/bias: operation 'VarHandleOp' is currently not implemented
dense_out/BiasAdd/ReadVariableOp: operation 'ReadVariableOp' is currently not implemented
```

With the exception of `StridedSlice` I can't find the other operations in the TF doc: https://www.tensorflow.org/api_docs/cc . Do you know if these are from some contrib packages?

Cheers,
- Lester

**Comment by lukealexmiller** [comment]
Is the above information logged out somewhere? I can't find the low-level info in vespa-logfmt. Would be good to be able to see this while debugging.

Testing a new model where I freeze and optimize the original model, and then remove embedding layers (my mistake including them in original model) I have the following file: https://drive.google.com/file/d/1fKQQwMmMl1RJu6W7gmXqtw2Yx1RxqRyh/view?usp=sharing 

However this gives the same error as before, even though all ops are supported in https://github.com/vespa-engine/vespa/blob/master/model-integration/src/main/java/ai/vespa/rankingexpression/importer/tensorflow/GraphImporter.java#L22 .

I then upgraded to latest Vespa (7.34.16 -> 7.62.30) and I now have a new error message:

Uploading application '/app/target/application.zip' using http://localhost:19071/application/v2/tenant/default/session
Session 2 for tenant 'default' created.
Preparing session 2 using http://localhost:19071/application/v2/tenant/default/session/2/prepared
Request failed. HTTP status code: 400
Invalid application package: default.default: Error loading model: Index 1 out of bounds for length 1

Tips on accessing the low-level information and any other insight would be greatly appreciated. Thanks!

**Comment by bratseth** [comment]
I reproduced this. The immediate cause is that MatMul is invoked with a rank 1 tensor, while it assumes 2 (i.e matrixes). I added an explicit check for that so the error message makes sense.

Now, why does this happen? That's where it gets tricky. You're right that we're missing a way to debug the intermediate graph created during conversion. I'm adding that now, not released yet but here's what it prints in this case:

dense_out/MatMul: 

MatMul(
  Join(
    Join(
	  ExpandDims(
	    Sum(
		  Join(
		    Join(
			  Argument(tensor(d0[],d1[32])), 
			  Map(
			    Join(
				  Sum(                                                 
				    Map(Argument(tensor(d0[],d1[32])), f(a)(a * a)),   
					Const(tensor()), 
					[input_embedding_user_guid_1]),  
				  Const(tensor()), 
				  f(a,b)(max(a, b))),                                
        		f(a)(1.0 / sqrt(a))),                                  
		      f(a,b)(a * b)), 
			  Join(
			    Argument(tensor(d0[],d1[32])), 
				Map(
				  Join(
				    Sum(
					  Map(
					    Argument(tensor(d0[],d1[32])), 
						f(a)(a * a)), 
					  Const(tensor()), 
					  [input_embedding_user_guid_1]), 
				  Const(tensor()), f(a,b)(max(a, b))), 
				  f(a)(1.0 / sqrt(a))), f(a,b)(a * b)), 
			  f(a,b)(a * b)), 
			Const(tensor()), 
			[input_embedding_user_guid_1]), 
		Const(tensor()), 
		[]), 
	  Join(
	    Map(
		  Join(
		    Const(tensor(test_batch_normalization_v1_moving_variance_0[1])), 
			Const(tensor()), 
			f(a,b)(a + b)), 
		  f(a)(1.0 / sqrt(a))), 
		Const(tensor(test_batch_normalization_v1_moving_variance_0[1])), 
		f(a,b)(a * b)), 
	  f(a,b)(a * b)), 
	Join(
	  Const(tensor(test_batch_normalization_v1_moving_mean_0[1])), 
	  Join(
	    Const(tensor(test_batch_normalization_v1_moving_mean_0[1])), 
		Join(
		  Map(
		    Join(
			  Const(tensor(test_batch_normalization_v1_moving_variance_0[1])), 
			  Const(tensor()), 
			  f(a,b)(a + b)), 
			f(a)(1.0 / sqrt(a))), 
		  Const(tensor(test_batch_normalization_v1_moving_variance_0[1])), 
		  f(a,b)(a * b)), f(a,b)(a * b)), 
		f(a,b)(a - b)), f(a,b)(a + b)), 
	  Const(tensor(test_dense_out_kernel_0[1],test_dense_out_kernel_1[1])))

Can't say I immediately see the cause - do you have a clue? Otherwise I think I'll need to add some deeper consistency checking at this level similar to what we have for our tensor language.

Btw, regarding the Keras/TF2.0 thing - we're aware and will add support for that.

**Comment by bratseth** [comment]
I added type annotations to the prettyprinting, which makes it easy to see the problem:

`
java.lang.IllegalArgumentException: Expected 2 dimensions in the first argument to MatMul(tensor(input_embedding_user_guid_0[]), tensor(test_dense_out_kernel_0[1],test_dense_out_kernel_1[1])) but got just [input_embedding_user_guid_0[]] from
tensor(input_embedding_user_guid_0[]):                                 Join(
tensor(input_embedding_user_guid_0[]):                                   Join(
tensor(input_embedding_user_guid_0[]):                                     ExpandDims(
tensor(input_embedding_user_guid_0[]):                                       Sum[keep_dims=false](
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[32]):         Join(
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[32]):           Join(
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[32]):             Argument(tensor(d0[],d1[32])),
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[1]):              Map(
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[1]):                Join(
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[1]):                  Sum[keep_dims=true](
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[32]):                   Map(
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[32]):                     Argument(tensor(d0[],d1[32])),
                                                                                           f(a)(a * a)
                                                                                         )
                                                                                         ,
tensor():                                                                                Const(tensor()),
                                                                                         [input_embedding_user_guid_1]
                                                                                       )
                                                                                       ,
tensor():                                                                              Const(tensor()), f(a,b)(max(a, b))
                                                                                     )
                                                                                     , f(a)(1.0 / sqrt(a))
                                                                                   )
                                                                                   , f(a,b)(a * b)
                                                                                 )
                                                                                 ,
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[32]):           Join(
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[32]):             Argument(tensor(d0[],d1[32])),
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[1]):              Map(
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[1]):                Join(
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[1]):                  Sum[keep_dims=true](
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[32]):                   Map(
tensor(input_embedding_user_guid_0[],input_embedding_user_guid_1[32]):                     Argument(tensor(d0[],d1[32])),
                                                                                           f(a)(a * a)
                                                                                         )
                                                                                         ,
tensor():                                                                                Const(tensor()),
                                                                                         [input_embedding_user_guid_1]
                                                                                       )
                                                                                       ,
tensor():                                                                              Const(tensor()), f(a,b)(max(a, b))
                                                                                     )
                                                                                     , f(a)(1.0 / sqrt(a))
                                                                                   )
                                                                                   , f(a,b)(a * b)
                                                                                 )
                                                                                 , f(a,b)(a * b)
                                                                               )
                                                                               ,
tensor():                                                                      Const(tensor()),
                                                                               [input_embedding_user_guid_1]
                                                                             )
                                                                             ,
tensor():                                                                    Const(tensor()),
                                                                             []
                                                                           )
                                                                           ,
tensor(test_batch_normalization_v1_moving_variance_0[1]):                  Join(
tensor(test_batch_normalization_v1_moving_variance_0[1]):                    Map(
tensor(test_batch_normalization_v1_moving_variance_0[1]):                      Join(
tensor(test_batch_normalization_v1_moving_variance_0[1]):                        Const(
                                                                                   tensor(
                                                                                     test_batch_normalization_v1_moving_variance_0[1]
                                                                                   )
                                                                                 )
                                                                                 ,
tensor():                                                                        Const(tensor()), f(a,b)(a + b)
                                                                               )
                                                                               , f(a)(1.0 / sqrt(a))
                                                                             )
                                                                             ,
tensor(test_batch_normalization_v1_moving_variance_0[1]):                    Const(
                                                                               tensor(
                                                                                 test_batch_normalization_v1_moving_variance_0[1]
                                                                               )
                                                                             )
                                                                             , f(a,b)(a * b)
                                                                           )
                                                                           , f(a,b)(a * b)
                                                                         )
                                                                         ,
tensor(test_batch_normalization_v1_moving_mean_0[1]):                    Join(
tensor(test_batch_normalization_v1_moving_mean_0[1]):                      Const(
                                                                             tensor(
                                                                               test_batch_normalization_v1_moving_mean_0[1]
                                                                             )
                                                                           )
                                                                           ,
tensor(test_batch_normalization_v1_moving_mean_0[1]):                      Join(
tensor(test_batch_normalization_v1_moving_mean_0[1]):                        Const(
                                                                               tensor(
                                                                                 test_batch_normalization_v1_moving_mean_0[1]
                                                                               )
                                                                             )
                                                                             ,
tensor(test_batch_normalization_v1_moving_variance_0[1]):                    Join(
tensor(test_batch_normalization_v1_moving_variance_0[1]):                      Map(
tensor(test_batch_normalization_v1_moving_variance_0[1]):                        Join(
tensor(test_batch_normalization_v1_moving_variance_0[1]):                          Const(
                                                                                     tensor(
                                                                                       test_batch_normalization_v1_moving_variance_0[1]
                                                                                     )
                                                                                   )
                                                                                   ,
tensor():                                                                          Const(tensor()), f(a,b)(a + b)
                                                                                 )
                                                                                 , f(a)(1.0 / sqrt(a))
                                                                               )
                                                                               ,
tensor(test_batch_normalization_v1_moving_variance_0[1]):                      Const(
                                                                                 tensor(
                                                                                   test_batch_normalization_v1_moving_variance_0[1]
                                                                                 )
                                                                               )
                                                                               , f(a,b)(a * b)
                                                                             )
                                                                             , f(a,b)(a * b)
                                                                           )
                                                                           , f(a,b)(a - b)
                                                                         )
                                                                         , f(a,b)(a + b)
                                                                       )
`

**Comment by bratseth** [comment]
The root cause was ExpandDims being unwilling to add a new dimension to the end of the list of dimensions (because it incorrectly ensured that the index of the new dimension is not larger than the existing set of dimensions). I fixed that, but there is another problem with this model which I'm looking at now.

**Comment by bratseth** [comment]
The other problem was that the intermediate data layout optimizer (how to lay out the data to cause the least jumping in memory to make the entire computation) could not find a solution for this model due to not being willing to work hard enough. All of this was fixed in https://github.com/vespa-engine/vespa/pull/9944 which is released 7.72.36.
